### ISSUE TITLE
New rule: module name should match module argument name

### ISSUE THREAD

**Issue by xman85** [cvisco/eslint-plugin-requirejs#96]
`define(["./cart", "./inventory"], function(cart, inventory) {`

Pretty self-explanatory: "./cart" -> cart, "./inventory" -> inventory.
The motivation is this: when I see "inventory" module variable in the code, I want to quickly navigate to the file where it's defined, and if filename doesn't match variable name it's annoying.

Thanks!

**Comment by platinumazure** [comment]
And what about paths involving multiple segments? Is the expected name just the last segment?

Personally I'd like to allow for case-insensitivity too, if I were to use this rule.

**Comment by cvisco** [comment]
@platinumazure yea I would see this as matching the `basename` of the path only

**Comment by xman85** [comment]
Yes, only the last segment - filename, because IDEs usually have a convenient search by filename.

**Comment by cvisco** [comment]
I'm curious about some common libs like:

``` js
define([ 'jquery', 'underscore' ], function ($, _) { /* ... */ });
```

Would you expect this rule to force that to be the following?

``` js
define([ 'jquery', 'underscore' ], function (jquery, underscore) { /* ... */ });
```

**Comment by cvisco** [comment]
Or would this apply only to relative paths (local modules)?

**Comment by xman85** [comment]
Good point, I think usually ppl are not interested in quickly navigating to a library source code. Moreover if a module has an alias, this feature doesn't make much sense because alias name can be different from filename. So only for local modules is sufficient.

**Comment by platinumazure** [comment]
Another possibility is to have an exceptions object:

``` json
{
    "rules": {
        "requirejs/module-match-file": ["error", { "exceptions": { "jquery": "$", "underscore": "_" } }]
    }
}
```

(I used a placeholder rule name, please pick a better one)

**Comment by cvisco** [comment]
@xman85 I understand your particular use case for this has to do with easier file navigation, but I see that as a side benefit of the rule itself. To me this is a much more general thing around enforcing naming conventions for module imports, so in that regard it needs to be sufficiently smart and/or flexible about what it warns on.

@platinumazure exceptions—or perhaps "aliases"—could work here, and i support the idea of a toggle for case sensitivity.

**Comment by cvisco** [comment]
I see one more potential issue here. What about filenames that would be illegal as identifiers in JavaScript? The modules in this plugin are a prime example of that: `libs/rules/no-assign-exports` for example.

**Comment by xman85** [comment]
If you filenames are not legal js identifiers then just don't enable module-match-file rule, I think :)

**Comment by cvisco** [comment]
Yea that's fair :)  

Would `consistent-dependency-names` be an acceptable name for this?

**Comment by xman85** [comment]
I liked "module-match-file" actually.

**Comment by cvisco** [comment]
Without context, though, that's too vague. And, it's not the module we're concerned with but the dependency list for the module.

How about `dependency-match-filename` ?

**Comment by xman85** [comment]
I'm okay with any name really, as long as it works. I thought "module" is an established require.js term that ppl understand.

**Comment by jDmacD** [comment]
Hi, this looks great and would make my boss deliriously happy. Are you far off implementing it?

**Comment by cvisco** [comment]
@jDmacD: Me personally? Yes. Unfortunately, due to other commitments, I wouldn't be able to schedule work on this until late December or January. If @xman85 or you or anyone else wants to take a stab at implementing it before then, feel free!

**Comment by alundiak** [comment]
Was this implemented since 2016?

**Comment by cvisco** [comment]
@alundiak no it hasn't been implemented yet
